### PR TITLE
MULE-11016: Race condition when running repeatable scheduled jobs

### PR DESCRIPTION
**Pull request by elrodro83**
* Fix after regression because of MULE-11339, add tests

### CODE REVIEW

File: services/scheduler/src/main/java/org/mule/service/scheduler/internal/AbstractRunnableFutureDecorator.java [hunk 1]
```diff
@@ -66,6 +69,35 @@ protected long beforeRun() {
     return startTime;
   }
 
+  protected void doRun(RunnableFuture<V> task, ClassLoader classLoader) {
```
Comment (pablokraan) [inline services/scheduler/src/main/java/org/mule/service/scheduler/internal/AbstractRunnableFutureDecorator.java]: javadoc?

File: services/scheduler/src/main/java/org/mule/service/scheduler/internal/AbstractRunnableFutureDecorator.java [hunk 2]
```diff
@@ -66,6 +69,35 @@ protected long beforeRun() {
...
+    final ClassLoader currentClassLoader = currentThread.getContextClassLoader();
+    currentThread.setContextClassLoader(classLoader);
+
+    try {
+      task.run();
+      task.get();
+    } catch (ExecutionException e) {
+      logger.error("Uncaught throwable in task " + toString(), e);
+    } catch (CancellationException e) {
+      logger.trace("Task " + toString() + " cancelled");
```
Comment (pablokraan) [inline services/scheduler/src/main/java/org/mule/service/scheduler/internal/AbstractRunnableFutureDecorator.java]: No need to re-throw the exceptions?

File: services/scheduler/src/main/java/org/mule/service/scheduler/internal/RunnableRepeatableFutureDecorator.java [hunk 3]
```diff
@@ -76,37 +74,28 @@ public void run() {
...
-      currentThread.setContextClassLoader(currentClassLoader);
-    }
+  @Override
+  protected long beforeRun() {
+    running = true;
+    return super.beforeRun();
   }
 
   @Override
   protected void wrapUp() {
```
Comment (pablokraan) [inline services/scheduler/src/main/java/org/mule/service/scheduler/internal/RunnableRepeatableFutureDecorator.java]: is there a missing throws exceptions in the method definition?

File: services/scheduler/src/main/java/org/mule/service/scheduler/internal/AbstractRunnableFutureDecorator.java (hunk 2 above)
Comment (elrodro83) [inline services/scheduler/src/main/java/org/mule/service/scheduler/internal/AbstractRunnableFutureDecorator.java]: No. That was just added no not clutter the log/console with cancellation stack traces. Reactor cancells jobs all the time and it fills the log

File: services/scheduler/src/main/java/org/mule/service/scheduler/internal/AbstractRunnableFutureDecorator.java [hunk 4]
```diff
@@ -66,7 +69,48 @@ protected long beforeRun() {
     return startTime;
   }
 
-  protected void wrapUp() {
+  /**
+   * Performas the required bookkeeping before and after running the task, as well as setting the appropriate context for the
```
Comment (pablokraan) [inline services/scheduler/src/main/java/org/mule/service/scheduler/internal/AbstractRunnableFutureDecorator.java]: Performas > Performs

**Pull request merged**